File: engine/src/router.rs
use crate::bridge::{BridgeAdapter, BridgeRegistry};
use crate::minimax::MinimaxSearcher;
use crate::path_discovery::{CandidatePath, PathDiscovery};
use crate::scoring::{compare_routes, get_strategy_weights, RouteScorer};
use crate::types::{
    BridgeQuote, Chain, Route, RouteHop, RouteRequest, RouterConfig, SearchStats, Strategy, Token,
};

/// The main MNMX cross-chain router.
pub struct MnmxRouter {
    config: RouterConfig,
    registry: BridgeRegistry,
}

/// Result of a routing operation.
#[derive(Debug)]
pub struct RoutingResult {
    pub best_route: Option<Route>,
    pub alternatives: Vec<Route>,
    pub stats: SearchStats,
}

impl MnmxRouter {
    /// Create a new router with the given configuration.
    pub fn new(config: RouterConfig) -> Self {
        Self {
            config,
            registry: BridgeRegistry::new(),
        }
    }

    /// Create a router with default config.
    pub fn default_router() -> Self {
        Self::new(RouterConfig::default())
    }

    /// Register a bridge adapter with the router.
    pub fn register_bridge(&mut self, adapter: Box<dyn BridgeAdapter>) {
        self.registry.register(adapter);
    }

    /// Set the bridge registry directly (useful for testing).
    pub fn set_registry(&mut self, registry: BridgeRegistry) {
        self.registry = registry;
    }

    /// Find the best route for the given request.
    pub fn find_route(&mut self, request: &RouteRequest) -> RoutingResult {
        // Update weights based on strategy
        self.config.weights = get_strategy_weights(request.strategy);
        self.config.max_hops = request.max_hops;
        self.config.strategy = request.strategy;

        // Run minimax search
        let mut searcher = MinimaxSearcher::new(self.config.clone());
        let (best_route, stats) = searcher.search(
            &self.registry,
            request.from_chain,
            &request.from_token,
            request.to_chain,
            &request.to_token,
            request.amount,
        );

        // Also find alternatives using different strategies
        let alternatives = self.find_alternative_routes(request, &best_route);

        RoutingResult {
            best_route,
            alternatives,
            stats,
        }
    }

    /// Find all candidate routes sorted by minimax score.
    pub fn find_all_routes(&mut self, request: &RouteRequest) -> (Vec<Route>, SearchStats) {
        let path_discovery = PathDiscovery::new(&self.registry, request.max_hops);
        let candidate_paths = path_discovery.discover_paths(
            request.from_chain,
            &request.from_token,
            request.to_chain,
            &request.to_token,
        );

        let scorer = RouteScorer::new(get_strategy_weights(request.strategy));
        let mut routes = Vec::new();

        for path in &candidate_paths {
            if let Some(route) = self.evaluate_candidate_path(path, request.amount) {
                routes.push(route);
            }
        }

        // Score all routes
        for route in &mut routes {
            route.minimax_score = scorer.score_route(route);
        }

        // Sort by minimax score (best first)
        routes.sort_by(compare_routes);

        let stats = SearchStats {
            nodes_explored: candidate_paths.len() as u64,
            nodes_pruned: 0,
            max_depth_reached: request.max_hops as u32,
            search_time_ms: 0,
        };

        (routes, stats)
    }

    /// Get all chains supported by registered bridges.
    pub fn get_supported_chains(&self) -> Vec<Chain> {
        let mut chains = std::collections::HashSet::new();
        for pair in self.registry.supported_pairs() {
            chains.insert(pair.0);
            chains.insert(pair.1);
        }
        let mut result: Vec<Chain> = chains.into_iter().collect();
        result.sort_by_key(|c| c.chain_id());
        result
    }

    /// Get bridge quotes for a specific chain pair and amount.
    pub fn get_quotes(
        &self,
        from_token: &Token,
        to_token: &Token,
        amount: f64,
    ) -> Vec<BridgeQuote> {
        let bridges = self
            .registry
            .get_bridges_for_pair(from_token.chain, to_token.chain);
        let mut quotes = Vec::new();
        for bridge in bridges {
            if let Some(quote) = bridge.get_quote(from_token, to_token, amount) {
                quotes.push(quote);
            }
        }
        // Sort by output amount (best first)
        quotes.sort_by(|a, b| {
            b.output_amount
                .partial_cmp(&a.output_amount)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        quotes
    }

    /// Update the router's configuration.
    pub fn update_config(&mut self, config: RouterConfig) {
        self.config = config;
    }

    /// Get current configuration.
    pub fn config(&self) -> &RouterConfig {
        &self.config
    }

    /// Get reference to the bridge registry.
    pub fn registry(&self) -> &BridgeRegistry {
        &self.registry
    }

    // ------- Internal methods -------

    /// Find alternative routes using strategies other than the primary one.
    fn find_alternative_routes(
        &self,
        request: &RouteRequest,
        primary: &Option<Route>,
    ) -> Vec<Route> {
        let alt_strategies = [Strategy::Cheapest, Strategy::Fastest, Strategy::Safest];
        let mut alternatives = Vec::new();

        let path_discovery = PathDiscovery::new(&self.registry, request.max_hops);
        let candidate_paths = path_discovery.discover_paths(
            request.from_chain,
            &request.from_token,
            request.to_chain,
            &request.to_token,
        );

        for strategy in &alt_strategies {
            if *strategy == request.strategy {
                continue;
            }

            let scorer = RouteScorer::new(get_strategy_weights(*strategy));
            let mut best_alt: Option<Route> = None;
            let mut best_score = f64::NEG_INFINITY;

            for path in &candidate_paths {
                if let Some(mut route) = self.evaluate_candidate_path(path, request.amount) {
                    let score = scorer.score_route(&route);
                    route.minimax_score = score;
                    if score > best_score {
                        best_score = score;
                        best_alt = Some(route);
                    }
                }
            }

            // Only include if different from primary
            if let Some(alt) = best_alt {
                let dominated_by_primary = match primary {
                    Some(p) => {
                        alt.hops.len() == p.hops.len()
                            && alt
                                .hops
                                .iter()
                                .zip(p.hops.iter())
                                .all(|(a, b)| a.bridge == b.bridge)
                    }
                    None => false,
                };
                if !dominated_by_primary {
                    alternatives.push(alt);
                }
            }
        }

        alternatives
    }

    /// Evaluate a candidate path by collecting quotes for each step and building a Route.
    fn evaluate_candidate_path(
        &self,
        path: &CandidatePath,
        initial_amount: f64,
    ) -> Option<Route> {
        let mut route = Route::new();
        let mut current_amount = initial_amount;

        for step in &path.steps {
            let bridges = self
                .registry
                .get_bridges_for_pair(step.from_chain, step.to_chain);

            // Find the specific bridge named in this step
            let bridge = bridges
                .iter()
                .find(|b| b.name() == step.bridge_name)?;

            let quote = bridge.get_quote(&step.from_token, &step.to_token, current_amount)?;

            route.hops.push(RouteHop {
                from_chain: step.from_chain,
                to_chain: step.to_chain,
                from_token: step.from_token.clone(),
                to_token: step.to_token.clone(),
                bridge: quote.bridge_name.clone(),
                input_amount: current_amount,
                output_amount: quote.output_amount,
                fee: quote.fee,
                estimated_time: quote.estimated_time,
            });

            route.total_fees += quote.fee;
            route.estimated_time += quote.estimated_time;
            current_amount = quote.output_amount;
        }

        route.expected_output = current_amount;
        route.guaranteed_minimum = current_amount * 0.95;

        Some(route)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::bridge::build_mock_registry;

    fn make_request(strategy: Strategy) -> RouteRequest {
        RouteRequest {
            from_chain: Chain::Ethereum,
            from_token: Token::new("USDC", Chain::Ethereum, 6, "0xaaa"),
            to_chain: Chain::Arbitrum,
            to_token: Token::new("USDC", Chain::Arbitrum, 6, "0xbbb"),
            amount: 10000.0,
            strategy,
            max_hops: 2,
        }
    }

    #[test]
    fn test_find_route() {
        let mut router = MnmxRouter::new(RouterConfig::default());
        router.set_registry(build_mock_registry());

        let request = make_request(Strategy::Minimax);
        let result = router.find_route(&request);
        assert!(result.best_route.is_some());
        let route = result.best_route.unwrap();
        assert!(!route.hops.is_empty());
        assert!(route.expected_output > 0.0);
        assert!(route.expected_output < 10000.0);
    }

    #[test]
    fn test_find_all_routes() {
        let mut router = MnmxRouter::new(RouterConfig::default());
        router.set_registry(build_mock_registry());

        let request = make_request(Strategy::Minimax);
        let (routes, _stats) = router.find_all_routes(&request);
        assert!(!routes.is_empty());

        // Should be sorted by score (descending)
        for w in routes.windows(2) {
            assert!(
                w[0].minimax_score >= w[1].minimax_score,
                "routes should be sorted by score"
            );
        }
    }

    #[test]
    fn test_get_quotes() {
        let mut router = MnmxRouter::new(RouterConfig::default());
        router.set_registry(build_mock_registry());

        let from = Token::new("USDC", Chain::Ethereum, 6, "0xaaa");
        let to = Token::new("USDC", Chain::Arbitrum, 6, "0xbbb");
        let quotes = router.get_quotes(&from, &to, 5000.0);
        assert!(!quotes.is_empty());
    }

    #[test]
    fn test_supported_chains() {
        let mut router = MnmxRouter::new(RouterConfig::default());
        router.set_registry(build_mock_registry());

        let chains = router.get_supported_chains();
        assert!(chains.len() >= 2);
    }

    #[test]
    fn test_different_strategies_may_differ() {
        let mut router = MnmxRouter::new(RouterConfig::default());
        router.set_registry(build_mock_registry());

        let request = make_request(Strategy::Minimax);
        let result = router.find_route(&request);
        assert!(result.best_route.is_some());
    }

    #[test]
    fn test_update_config() {
        let mut router = MnmxRouter::new(RouterConfig::default());
        let new_config = RouterConfig {
            strategy: Strategy::Safest,
            max_hops: 1,
            ..RouterConfig::default()
        };
        router.update_config(new_config);
        assert_eq!(router.config().strategy, Strategy::Safest);
        assert_eq!(router.config().max_hops, 1);
    }
}

File: sdk/python/mnmx/router.py
"""Core MNMX router: minimax-based cross-chain path discovery."""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from itertools import permutations
from typing import Any

from mnmx.bridges import BridgeAdapter, BridgeRegistry, create_default_registry
from mnmx.exceptions import (
    InvalidConfigError,
    NoRouteFoundError,
    RouteTimeoutError,
)
from mnmx.scoring import RouteScorer, get_strategy_weights
from mnmx.types import (
    AdversarialModel,
    BridgeQuote,
    Chain,
    Route,
    RouteHop,
    RouteRequest,
    RouterConfig,
    ScoringWeights,
    SearchStats,
    Strategy,
    VALID_STRATEGIES,
)


@dataclass
class _SearchNode:
    """Internal node in the minimax game tree."""

    chain: Chain
    token: str
    amount: float
    depth: int
    hops: list[RouteHop] = field(default_factory=list)
    total_fee: float = 0.0
    total_time: int = 0


class MnmxRouter:
    """Cross-chain router using minimax search with alpha-beta pruning.

    The router models cross-chain routing as a two-player game:
    - MAX player: the user, choosing the best bridge at each hop
    - MIN player: the adversarial market (slippage, MEV, delays)

    The minimax search finds the route whose *worst-case* outcome
    is maximised (the guaranteed minimum).
    """

    def __init__(
        self,
        strategy: Strategy = "minimax",
        config: RouterConfig | None = None,
        registry: BridgeRegistry | None = None,
        **kwargs: Any,
    ) -> None:
        if config is not None:
            self._config = config
        else:
            weights = kwargs.get("weights")
            adversarial = kwargs.get("adversarial_model")
            self._config = RouterConfig(
                strategy=strategy,
                slippage_tolerance=kwargs.get("slippage_tolerance", 0.005),
                timeout_ms=kwargs.get("timeout_ms", 5000),
                max_hops=kwargs.get("max_hops", 3),
                weights=weights if isinstance(weights, ScoringWeights) else ScoringWeights(),
                adversarial_model=adversarial if isinstance(adversarial, AdversarialModel) else AdversarialModel(),
            )
        self._registry = registry or create_default_registry()
        self._scorer = RouteScorer(self._config.weights)
        self._stats = SearchStats(0, 0, 0, 0.0)

    # ---- public API --------------------------------------------------------

    @property
    def config(self) -> RouterConfig:
        return self._config

    @property
    def last_search_stats(self) -> SearchStats:
        return self._stats

    def find_route(
        self,
        from_chain: str | Chain,
        from_token: str,
        amount: float,
        to_chain: str | Chain,
        to_token: str,
        **kwargs: Any,
    ) -> Route:
        """Find the single best route using the configured strategy."""
        routes = self.find_all_routes(from_chain, from_token, amount, to_chain, to_token, **kwargs)
        if not routes:
            src = from_chain if isinstance(from_chain, str) else from_chain.value
            dst = to_chain if isinstance(to_chain, str) else to_chain.value
            raise NoRouteFoundError(src, dst, from_token, to_token)
        return routes[0]

    def find_all_routes(
        self,
        from_chain: str | Chain,
        from_token: str,
        amount: float,
        to_chain: str | Chain,
        to_token: str,
        **kwargs: Any,
    ) -> list[Route]:
        """Find all viable routes, sorted best-first by minimax score."""
        request = self._build_request(from_chain, from_token, amount, to_chain, to_token, **kwargs)
        self._validate_request(request)

        start_ms = time.monotonic() * 1000

        # discover candidate paths (sequences of (chain, bridge) stops)
        candidate_paths = self._discover_paths(request)

        # run minimax on each path to get scored routes
        routes: list[Route] = []
        for path_chains, path_bridges in candidate_paths:
            elapsed = time.monotonic() * 1000 - start_ms
            if elapsed > self._config.timeout_ms:
                break
            route = self._evaluate_path(request, path_chains, path_bridges)
            if route is not None:
                routes.append(route)

        self._stats.search_time_ms = time.monotonic() * 1000 - start_ms

        # sort by minimax score descending
        strategy = kwargs.get("strategy", self._config.strategy)
        weights = get_strategy_weights(strategy)
        for r in routes:
            r.minimax_score = self._scorer.score_route(r, weights)
            r.strategy = strategy

        routes.sort(key=lambda r: r.minimax_score, reverse=True)
        return routes

    def get_supported_chains(self) -> list[str]:
        return Chain.all_names()

    def get_supported_bridges(self) -> list[str]:
        return self._registry.names()

    # ---- internal ----------------------------------------------------------

    def _build_request(
        self,
        from_chain: str | Chain,
        from_token: str,
        amount: float,
        to_chain: str | Chain,
        to_token: str,
        **kwargs: Any,
    ) -> RouteRequest:
        src = Chain.from_str(from_chain) if isinstance(from_chain, str) else from_chain
        dst = Chain.from_str(to_chain) if isinstance(to_chain, str) else to_chain
        return RouteRequest(
            from_chain=src,
            from_token=from_token,
            amount=amount,
            to_chain=dst,
            to_token=to_token,
            strategy=kwargs.get("strategy", self._config.strategy),
            max_hops=kwargs.get("max_hops", self._config.max_hops),
            slippage_tolerance=kwargs.get("slippage_tolerance", self._config.slippage_tolerance),
        )

    def _validate_request(self, request: RouteRequest) -> None:
        if request.from_chain == request.to_chain and request.from_token == request.to_token:
            raise InvalidConfigError("route", "Source and destination are identical")
        if request.amount <= 0:
            raise InvalidConfigError("amount", "Amount must be positive")

    def _discover_paths(
        self, request: RouteRequest
    ) -> list[tuple[list[Chain], list[str]]]:
        """Enumerate candidate paths as (chain sequence, bridge sequence)."""
        results: list[tuple[list[Chain], list[str]]] = []

        src = request.from_chain
        dst = request.to_chain

        # 1-hop direct routes
        direct_bridges = self._registry.get_for_pair(src, dst)
        for bridge in direct_bridges:
            results.append(([src, dst], [bridge.name]))

        if request.max_hops < 2:
            return results

        # 2-hop routes via intermediate chains
        all_chains = list(Chain)
        intermediate_chains = [c for c in all_chains if c != src and c != dst]
        for mid in intermediate_chains:
            bridges_leg1 = self._registry.get_for_pair(src, mid)
            bridges_leg2 = self._registry.get_for_pair(mid, dst)
            for b1 in bridges_leg1:
                for b2 in bridges_leg2:
                    results.append(([src, mid, dst], [b1.name, b2.name]))

        if request.max_hops < 3:
            return results

        # 3-hop routes via two intermediates
        for mid1 in intermediate_chains:
            for mid2 in intermediate_chains:
                if mid1 == mid2:
                    continue
                b1s = self._registry.get_for_pair(src, mid1)
                b2s = self._registry.get_for_pair(mid1, mid2)
                b3s = self._registry.get_for_pair(mid2, dst)
                for b1 in b1s:
                    for b2 in b2s:
                        for b3 in b3s:
                            results.append(
                                ([src, mid1, mid2, dst], [b1.name, b2.name, b3.name])
                            )
                            # limit combinatorial explosion
                            if len(results) > 500:
                                return results

        return results

    def _evaluate_path(
        self,
        request: RouteRequest,
        chain_sequence: list[Chain],
        bridge_names: list[str],
    ) -> Route | None:
        """Run minimax evaluation on a single candidate path."""
        self._stats.nodes_explored += 1

        hops: list[RouteHop] = []
        current_amount = request.amount
        total_fee = 0.0
        total_time = 0
        from_token = request.from_token

        for i, bridge_name in enumerate(bridge_names):
            src_chain = chain_sequence[i]
            dst_chain = chain_sequence[i + 1]
            to_token = request.to_token if i == len(bridge_names) - 1 else from_token

            try:
                bridge = self._registry.get(bridge_name)
                quote = bridge.get_quote(src_chain, dst_chain, from_token, to_token, current_amount)
            except Exception:
                self._stats.nodes_pruned += 1
                return None

            # adversarial adjustment (MIN player moves)
            adv = self._config.adversarial_model
            adversarial_fee = quote.fee * adv.gas_multiplier
            adversarial_slippage = quote.slippage * adv.slippage_multiplier
            mev_loss = current_amount * adv.mev_extraction
            price_impact = current_amount * adv.price_movement

            worst_case_output = (
                current_amount
                - adversarial_fee
                - (current_amount * adversarial_slippage)
                - mev_loss
                - price_impact
            )
            worst_case_output = max(worst_case_output, 0.0)

            hop = RouteHop(
                from_chain=src_chain,
                to_chain=dst_chain,
                from_token=from_token,
                to_token=to_token,
                bridge=bridge_name,
                input_amount=current_amount,
                output_amount=quote.output_amount,
                fee=quote.fee,
                estimated_time=int(quote.estimated_time * adv.bridge_delay_multiplier),
            )
            hops.append(hop)

            total_fee += quote.fee
            total_time += hop.estimated_time
            current_amount = quote.output_amount
            from_token = to_token

        if not hops:
            return None

        self._stats.max_depth_reached = max(self._stats.max_depth_reached, len(hops))

        # guaranteed minimum via minimax: apply adversarial model to expected output
        expected_output = current_amount
        guaranteed_minimum = self._compute_guaranteed_minimum(hops, request.amount)

        route = Route(
            path=hops,
            expected_output=expected_output,
            guaranteed_minimum=guaranteed_minimum,
            total_fees=total_fee,
            estimated_time=total_time,
            minimax_score=0.0,  # scored later
            strategy=request.strategy,
        )
        return route

    def _compute_guaranteed_minimum(self, hops: list[RouteHop], initial_amount: float) -> float:
        """Compute worst-case guaranteed output using adversarial model."""
        adv = self._config.adversarial_model
        amount = initial_amount

        for hop in hops:
            fee_worst = hop.fee * adv.gas_multiplier
            slippage_worst = amount * (hop.fee / hop.input_amount) * adv.slippage_multiplier
            mev = amount * adv.mev_extraction
            price = amount * adv.price_movement
            amount = amount - fee_worst - slippage_worst - mev - price
            amount = max(amount, 0.0)

        return amount

    def _run_minimax(
        self,
        node: _SearchNode,
        target_chain: Chain,
        target_token: str,
        alpha: float,
        beta: float,
        is_maximizing: bool,
        request: RouteRequest,
    ) -> float:
        """Recursive minimax with alpha-beta pruning.

        MAX player picks the best bridge.
        MIN player applies worst-case adversarial conditions.
        """
        self._stats.nodes_explored += 1

        # terminal: reached destination
        if node.chain == target_chain and node.depth > 0:
            return node.amount

        # depth limit
        if node.depth >= request.max_hops:
            self._stats.nodes_pruned += 1
            return 0.0 if node.chain != target_chain else node.amount

        bridges = self._registry.get_for_pair(node.chain, target_chain)
        if not bridges and node.depth < request.max_hops - 1:
            # try intermediate hops
            all_adapters = self._registry.get_all()
            reachable: list[tuple[BridgeAdapter, Chain]] = []
            for adapter in all_adapters:
                for chain in adapter.supported_chains:
                    if chain != node.chain and adapter.supports_pair(node.chain, chain):
                        reachable.append((adapter, chain))
            if not reachable:
                return 0.0
            bridges_with_targets = reachable
        else:
            bridges_with_targets = [(b, target_chain) for b in bridges]

        if is_maximizing:
            best = float("-inf")
            for bridge, next_chain in bridges_with_targets:
                try:
                    quote = bridge.get_quote(
                        node.chain, next_chain, node.token, target_token, node.amount
                    )
                except Exception:
                    self._stats.nodes_pruned += 1
                    continue

                child = _SearchNode(
                    chain=next_chain,
                    token=target_token,
                    amount=quote.output_amount,
                    depth=node.depth + 1,
                )
                val = self._run_minimax(child, target_chain, target_token, alpha, beta, False, request)
                best = max(best, val)
                alpha = max(alpha, val)
                if beta <= alpha:
                    self._stats.nodes_pruned += 1
                    break
            return best if best != float("-inf") else 0.0
        else:
            # MIN player: apply adversarial conditions
            adv = self._config.adversarial_model
            worst_amount = node.amount * (1.0 - adv.mev_extraction - adv.price_movement)
            worst_amount = max(worst_amount, 0.0)
            child = _SearchNode(
                chain=node.chain,
                token=node.token,
                amount=worst_amount,
                depth=node.depth,
            )
            return self._run_minimax(child, target_chain, target_token, alpha, beta, True, request)

    def _collect_quotes(
        self,
        from_chain: Chain,
        to_chain: Chain,
        from_token: str,
        to_token: str,
        amount: float,
    ) -> list[BridgeQuote]:
        """Collect quotes from all bridges supporting the given pair."""
        bridges = self._registry.get_for_pair(from_chain, to_chain)
        quotes: list[BridgeQuote] = []
        for bridge in bridges:
            try:
                q = bridge.get_quote(from_chain, to_chain, from_token, to_token, amount)
                quotes.append(q)
            except Exception:
                continue
        return quotes

    def _apply_strategy_weights(self, routes: list[Route], strategy: Strategy) -> list[Route]:
        """Re-score and sort routes according to strategy weights."""
        weights = get_strategy_weights(strategy)
        for route in routes:
            route.minimax_score = self._scorer.score_route(route, weights)
            route.strategy = strategy
        routes.sort(key=lambda r: r.minimax_score, reverse=True)
        return routes
